server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java

```diff
@@ -0,0 +1,376 @@
+    private final AckListener ackListener;
+    private final LongSupplier currentTimeSupplier;
+    private final long startTime;
+
+    private Optional<ApplyCommitRequest> applyCommitRequest; // set when state is committed
+    private boolean isCompleted; // set when publication is completed
+    private boolean timedOut; // set when publication timed out
+
+    public Publication(Settings settings, PublishRequest publishRequest, AckListener ackListener, LongSupplier currentTimeSupplier) {
+        super(settings);
+        this.publishRequest = publishRequest;
+        this.ackListener = ackListener;
+        this.currentTimeSupplier = currentTimeSupplier;
+        startTime = currentTimeSupplier.getAsLong();
+        applyCommitRequest = Optional.empty();
+        publicationTargets = new ArrayList<>(publishRequest.getAcceptedState().getNodes().getNodes().size());
+        publishRequest.getAcceptedState().getNodes().iterator().forEachRemaining(n -> publicationTargets.add(new PublicationTarget(n)));
+    }
+
+    public void start(Set<DiscoveryNode> faultyNodes) {
+        logger.trace("publishing {} to {}", publishRequest, publicationTargets);
+
+        for (final DiscoveryNode faultyNode : faultyNodes) {
+            onFaultyNode(faultyNode);
+        }
+        onPossibleCommitFailure();
+        publicationTargets.forEach(PublicationTarget::sendPublishRequest);
+    }
+
+    public void onTimeout() {
+        assert timedOut == false;
+        timedOut = true;
+        if (applyCommitRequest.isPresent() == false) {
+            logger.debug("onTimeout: [{}] timed out before committing", this);
+            // fail all current publications
+            final Exception e = new ElasticsearchException("publication timed out before committing");
+            publicationTargets.stream().filter(PublicationTarget::isActive).forEach(pt -> pt.setFailed(e));
+        }
+        onPossibleCompletion();
+    }
+
+    public void onFaultyNode(DiscoveryNode faultyNode) {
+        publicationTargets.forEach(t -> t.onFaultyNode(faultyNode));
+        onPossibleCompletion();
+    }
+
+    private void onPossibleCompletion() {
+        if (isCompleted) {
+            return;
+        }
+
+        if (timedOut == false) {
+            for (final PublicationTarget target : publicationTargets) {
+                if (target.publicationTargetStateMachine.isActive()) {
+                    return;
+                }
+            }
+        }
+
+        if (applyCommitRequest.isPresent() == false) {
+            logger.debug("onPossibleCompletion: [{}] commit failed", this);
+            assert isCompleted == false;
+            isCompleted = true;
+            onCompletion(false);
+            return;
+        }
+
+        assert isCompleted == false;
+        isCompleted = true;
+        onCompletion(true);
+        assert applyCommitRequest.isPresent();
+        logger.trace("onPossibleCompletion: [{}] was successful, applying new state locally", this);
```

Comment:
not sure it makes sense to talk about `applying` the state here

Reply:
fixed in 25d5835ea8c